Core/NoOpRunner.Core/Shapes/BaseShape.cs

```diff
@@ -24,22 +29,25 @@ public BaseShape(GenerationStrategy strategy, int lowerBoundX, int lowerBoundY,
             var blocks = strategy.GenerateShapeBlocks(lowerBoundX, lowerBoundY, upperBoundX, upperBoundY);
             CenterPosX = blocks[0].OffsetX;
             CenterPosY = blocks[0].OffsetY;
+            this.lowerBoundY = lowerBoundY;
+            this.upperBoundY = upperBoundY;
             ShapeBlocks = GenerationStrategy.MakeRelative(blocks, CenterPosX, CenterPosY);
+            Strategy = strategy;
         }
 
         public (int[], int[]) GetCoords()
         {
             List<int> xCoords = new List<int>();
             List<int> yCoords = new List<int>();
 
-            ShapeBlocks.ForEach(x =>
```

Comment:
Why change into foreach?

Reply:
Because unlike `ShapeBlocks`, `VisibleShapeBlocks` is an `IEnumerable` (I see no reason for it to be a list), and using `ForEach` is smelly anyways (unlike linq functions, it only exists for lists and modifies the list in-place. `foreach` is preferrable)